MySQL mode `show columns from table`, if `modificationMetaId` changed between prepared and execute. Then error eccored.

I found these code in Prepared
```
    public boolean needRecompile() {
        Database db = session.getDatabase();
        if (db == null) {
            throw DbException.get(ErrorCode.CONNECTION_BROKEN_1, "database closed");
        }
        // parser: currently, compiling every create/drop/... twice
        // because needRecompile return true even for the first execution
        return prepareAlways ||
                modificationMetaId < db.getModificationMetaId() ||
                db.getSettings().recompileAlways;
    }
```
if `modificationMetaId` changed between prepared and execute. 
`show columns from table ` will change to 
```
SELECT C.COLUMN_NAME                                    FIELD,
       DATA_TYPE_SQL(?2, ?1, 'TABLE', C.DTD_IDENTIFIER) TYPE,
       C.IS_NULLABLE                                    "NULL",
       CASE (SELECT MAX(I.INDEX_TYPE_NAME)
             FROM INFORMATION_SCHEMA.INDEXES I
                      JOIN INFORMATION_SCHEMA.INDEX_COLUMNS IC
             WHERE I.TABLE_SCHEMA = C.TABLE_SCHEMA
               AND I.TABLE_NAME = C.TABLE_NAME
               AND IC.TABLE_SCHEMA = C.TABLE_SCHEMA
               AND IC.TABLE_NAME = C.TABLE_NAME
               AND IC.INDEX_SCHEMA = I.INDEX_SCHEMA
               AND IC.INDEX_NAME = I.INDEX_NAME
               AND IC.COLUMN_NAME = C.COLUMN_NAME)
           WHEN 'PRIMARY KEY' THEN 'PRI'
           WHEN 'UNIQUE INDEX' THEN 'UNI'
           ELSE '' END                                  `KEY`,
       COALESCE(COLUMN_DEFAULT, 'NULL')                 `DEFAULT`
FROM INFORMATION_SCHEMA.COLUMNS C
WHERE C.TABLE_NAME = ?1
  AND C.TABLE_SCHEMA = ?2
ORDER BY C.ORDINAL_POSITION
```
 but recompile has no parameters. Then errors occured like this.
```
Caused by: java.lang.IndexOutOfBoundsException: Index: 1, Size: 0
 at java.util.ArrayList.rangeCheck(ArrayList.java:659)
 at java.util.ArrayList.get(ArrayList.java:435)
 at org.h2.command.Parser.readParameter(Parser.java:4877)
 at org.h2.command.Parser.readTermWithoutIdentifier(Parser.java:4950)
 at org.h2.command.Parser.readTerm(Parser.java:4901)
 at org.h2.command.Parser.readFactor(Parser.java:3398)
 at org.h2.command.Parser.readSum(Parser.java:3385)
 at org.h2.command.Parser.readConcat(Parser.java:3350)
 at org.h2.command.Parser.readCondition(Parser.java:3132)
 at org.h2.command.Parser.readExpression(Parser.java:3053)
 at org.h2.command.Parser.readBuiltinFunctionIf(Parser.java:4391)
 at org.h2.command.Parser.readFunction(Parser.java:3829)
 at org.h2.command.Parser.readTermWithIdentifier(Parser.java:5211)
 at org.h2.command.Parser.readTerm(Parser.java:4901)
 at org.h2.command.Parser.readFactor(Parser.java:3398)
 at org.h2.command.Parser.readSum(Parser.java:3385)
 at org.h2.command.Parser.readConcat(Parser.java:3350)
 at org.h2.command.Parser.readCondition(Parser.java:3132)
 at org.h2.command.Parser.readExpression(Parser.java:3053)
 at org.h2.command.Parser.parseSelectExpressions(Parser.java:2853)
 at org.h2.command.Parser.parseSelect(Parser.java:2871)
 at org.h2.command.Parser.parseQueryPrimary(Parser.java:2762)
 at org.h2.command.Parser.parseQueryTerm(Parser.java:2633)
 at org.h2.command.Parser.parseQueryExpressionBody(Parser.java:2612)
 at org.h2.command.Parser.parseQueryExpressionBodyAndEndOfQuery(Parser.java:2605)
 at org.h2.command.Parser.parseQueryExpression(Parser.java:2598)
 at org.h2.command.Parser.parseQuery(Parser.java:2567)
 at org.h2.command.Parser.parsePrepared(Parser.java:724)
 at org.h2.command.Parser.parse(Parser.java:689)
 at org.h2.command.Parser.parse(Parser.java:661)
 at org.h2.command.CommandContainer.recompileIfRequired(CommandContainer.java:146)
 at org.h2.command.CommandContainer.query(CommandContainer.java:247)
 at org.h2.command.Command.executeQuery(Command.java:190)
```

--- a/h2/src/docsrc/html/changelog.html
+++ b/h2/src/docsrc/html/changelog.html
@@ -21,6 +21,8 @@ Change Log
 
 <h2>Next Version (unreleased)</h2>
 <ul>
+<li>Issue #3701: CLOBs can cause ClassCastExceptions
+</li>
 <li>Issue #3698: MySQL mode show columns from table, if modificationMetaId changed between prepared and execute.
 Then error occurred.
 </li>

--- a/h2/src/docsrc/html/changelog.html
+++ b/h2/src/docsrc/html/changelog.html
@@ -21,6 +21,11 @@ Change Log
 
 <h2>Next Version (unreleased)</h2>
 <ul>
+<li>Issue #3698: MySQL mode show columns from table, if modificationMetaId changed between prepared and execute.
+Then error occurred.
+</li>
+<li>PR #3699: Upgrade to the latest OSGi JDBC specification
+</li>
 <li>Issue #3659: User-defined variable "sticky" if used in view with join
 </li>
 <li>Issue #3693: Wrong result when intersecting unnested arrays

--- a/h2/src/main/org/h2/command/Parser.java
+++ b/h2/src/main/org/h2/command/Parser.java
@@ -1317,7 +1317,6 @@ public class Parser {
     }
 
     private Prepared parseShow() {
-        ArrayList<Value> paramValues = Utils.newSmallArrayList();
         StringBuilder buff = new StringBuilder("SELECT ");
         if (readIf("CLIENT_ENCODING")) {
             // for PostgreSQL compatibility
@@ -1365,22 +1364,25 @@ public class Parser {
             }
             buff.append("TABLE_NAME, TABLE_SCHEMA FROM "
                     + "INFORMATION_SCHEMA.TABLES "
-                    + "WHERE TABLE_SCHEMA=? ORDER BY TABLE_NAME");
-            paramValues.add(ValueVarchar.get(schema));
+                    + "WHERE TABLE_SCHEMA=");
+            StringUtils.quoteStringSQL(buff, schema).append(" ORDER BY TABLE_NAME");
         } else if (readIf("COLUMNS")) {
             // for MySQL compatibility
             read(FROM);
             String tableName = readIdentifierWithSchema();
             String schemaName = getSchema().getName();
-            paramValues.add(ValueVarchar.get(tableName));
             if (readIf(FROM)) {
                 schemaName = readIdentifier();
             }
             buff.append("C.COLUMN_NAME FIELD, ");
             boolean oldInformationSchema = session.isOldInformationSchema();
-            buff.append(oldInformationSchema
-                    ? "C.COLUMN_TYPE"
-                    : "DATA_TYPE_SQL(?2, ?1, 'TABLE', C.DTD_IDENTIFIER)");
+            if (oldInformationSchema) {
+                buff.append("C.COLUMN_TYPE");
+            } else {
+                buff.append("DATA_TYPE_SQL(");
+                StringUtils.quoteStringSQL(buff, schemaName).append(", ");
+                StringUtils.quoteStringSQL(buff, tableName).append(", 'TABLE', C.DTD_IDENTIFIER)");
+            }
             buff.append(" TYPE, "
                     + "C.IS_NULLABLE \"NULL\", "
                     + "CASE (SELECT MAX(I.INDEX_TYPE_NAME) FROM "
@@ -1404,9 +1406,9 @@ public class Parser {
                     + "WHEN 'UNIQUE INDEX' THEN 'UNI' ELSE '' END `KEY`, "
                     + "COALESCE(COLUMN_DEFAULT, 'NULL') `DEFAULT` "
                     + "FROM INFORMATION_SCHEMA.COLUMNS C "
-                    + "WHERE C.TABLE_NAME=?1 AND C.TABLE_SCHEMA=?2 "
-                    + "ORDER BY C.ORDINAL_POSITION");
-            paramValues.add(ValueVarchar.get(schemaName));
+                    + "WHERE C.TABLE_SCHEMA=");
+            StringUtils.quoteStringSQL(buff, schemaName).append(" AND C.TABLE_NAME=");
+            StringUtils.quoteStringSQL(buff, tableName).append(" ORDER BY C.ORDINAL_POSITION");
         } else if (readIf("DATABASES") || readIf("SCHEMAS")) {
             // for MySQL compatibility
             buff.append("SCHEMA_NAME FROM INFORMATION_SCHEMA.SCHEMATA");
@@ -1416,26 +1418,12 @@ public class Parser {
         }
         boolean b = session.getAllowLiterals();
         try {
-            // need to temporarily enable it, in case we are in
-            // ALLOW_LITERALS_NUMBERS mode
+            // need to temporarily enable it
             session.setAllowLiterals(true);
-            return prepare(session, buff.toString(), paramValues);
+            return session.prepare(buff.toString());
         } finally {
             session.setAllowLiterals(b);
         }
-    }
-
-    private static Prepared prepare(SessionLocal s, String sql,
-            ArrayList<Value> paramValues) {
-        Prepared prep = s.prepare(sql);
-        ArrayList<Parameter> params = prep.getParameters();
-        if (params != null) {
-            for (int i = 0, size = params.size(); i < size; i++) {
-                Parameter p = params.get(i);
-                p.setValue(paramValues.get(i));
-            }
-        }
-        return prep;
     }
 
     private boolean isDerivedTable() {

--- a/h2/src/main/org/h2/value/ValueBlob.java
+++ b/h2/src/main/org/h2/value/ValueBlob.java
@@ -236,9 +236,7 @@ public final class ValueBlob extends ValueLob {
         if ((sqlFlags & REPLACE_LOBS_FOR_TRACE) != 0
                 && (!(lobData instanceof LobDataInMemory) || octetLength > SysProperties.MAX_TRACE_DATA_LENGTH)) {
             builder.append("CAST(REPEAT(CHAR(0), ").append(octetLength).append(") AS BINARY VARYING");
-            LobDataDatabase lobDb = (LobDataDatabase) lobData;
-            builder.append(" /* table: ").append(lobDb.getTableId()).append(" id: ").append(lobDb.getLobId())
-                    .append(" */)");
+            formatLobDataComment(builder);
         } else {
             if ((sqlFlags & (REPLACE_LOBS_FOR_TRACE | NO_CASTS)) == 0) {
                 builder.append("CAST(X'");

--- a/h2/src/main/org/h2/value/ValueClob.java
+++ b/h2/src/main/org/h2/value/ValueClob.java
@@ -278,9 +278,7 @@ public final class ValueClob extends ValueLob {
         if ((sqlFlags & REPLACE_LOBS_FOR_TRACE) != 0
                 && (!(lobData instanceof LobDataInMemory) || charLength > SysProperties.MAX_TRACE_DATA_LENGTH)) {
             builder.append("SPACE(").append(charLength);
-            LobDataDatabase lobDb = (LobDataDatabase) lobData;
-            builder.append(" /* table: ").append(lobDb.getTableId()).append(" id: ").append(lobDb.getLobId())
-                    .append(" */)");
+            formatLobDataComment(builder);
         } else {
             if ((sqlFlags & (REPLACE_LOBS_FOR_TRACE | NO_CASTS)) == 0) {
                 StringUtils.quoteStringSQL(builder.append("CAST("), getString()).append(" AS CHARACTER LARGE OBJECT(")

--- a/h2/src/main/org/h2/value/ValueLob.java
+++ b/h2/src/main/org/h2/value/ValueLob.java
@@ -24,6 +24,7 @@ import org.h2.util.StringUtils;
 import org.h2.util.Utils;
 import org.h2.value.lob.LobData;
 import org.h2.value.lob.LobDataDatabase;
+import org.h2.value.lob.LobDataFetchOnDemand;
 import org.h2.value.lob.LobDataInMemory;
 
 /**
@@ -292,6 +293,21 @@ public abstract class ValueLob extends Value {
             }
         }
         return this;
+    }
+
+    final void formatLobDataComment(StringBuilder builder) {
+        if (lobData instanceof LobDataDatabase) {
+            LobDataDatabase lobDb = (LobDataDatabase) lobData;
+            builder.append(" /* table: ").append(lobDb.getTableId()).append(" id: ").append(lobDb.getLobId())
+                    .append(" */)");
+        } else if (lobData instanceof LobDataFetchOnDemand) {
+            LobDataFetchOnDemand lobDemand = (LobDataFetchOnDemand) lobData;
+            builder.append(" /* table: ").append(lobDemand.getTableId()).append(" id: ")
+                    .append(lobDemand.getLobId()).append(" */)");
+        } else {
+            builder.append(" /* ").append(lobData.toString().replaceAll("\\*/", "\\\\*\\\\/"))
+                    .append(" */");
+        }
     }
 
 }

--- a/h2/src/test/org/h2/test/db/TestCases.java
+++ b/h2/src/test/org/h2/test/db/TestCases.java
@@ -83,6 +83,7 @@ public class TestCases extends TestDb {
         testExplainAnalyze();
         testDataChangeDeltaTable();
         testGroupSortedReset();
+        testShowColumns();
         if (config.memory) {
             return;
         }
@@ -1764,6 +1765,24 @@ public class TestCases extends TestDb {
         stat.execute(sql);
         stat.execute("UPDATE T1 SET B = 7 WHERE A = 3");
         stat.execute(sql);
+        conn.close();
+    }
+
+    private void testShowColumns() throws SQLException {
+        // This test requires a PreparedStatement
+        deleteDb("cases");
+        Connection conn = getConnection("cases");
+        Statement stat = conn.createStatement();
+        stat.execute("CREATE TABLE TEST(A INTEGER)");
+        PreparedStatement prep = conn.prepareStatement("SHOW COLUMNS FROM TEST");
+        ResultSet rs = prep.executeQuery();
+        assertTrue(rs.next());
+        assertFalse(rs.next());
+        stat.execute("ALTER TABLE TEST ADD COLUMN B INTEGER");
+        rs = prep.executeQuery();
+        assertTrue(rs.next());
+        assertTrue(rs.next());
+        assertFalse(rs.next());
         conn.close();
     }
 

--- a/h2/src/test/org/h2/test/jdbc/TestLobApi.java
+++ b/h2/src/test/org/h2/test/jdbc/TestLobApi.java
@@ -278,6 +278,7 @@ public class TestLobApi extends TestDb {
         rs = stat.executeQuery("select * from test");
         rs.next();
         Blob b3 = rs.getBlob(2);
+        b3.toString();
         assertEquals(length, b3.length());
         byte[] bytes = b.getBytes(1, length);
         byte[] bytes2 = b3.getBytes(1, length);
@@ -370,6 +371,7 @@ public class TestLobApi extends TestDb {
         rs = stat.executeQuery("select * from test");
         rs.next();
         Clob c2 = rs.getClob(2);
+        c2.toString();
         assertEquals(length, c2.length());
         String s = c.getSubString(1, length);
         String s2 = c2.getSubString(1, length);

--- a/h2/src/test/org/h2/test/store/TestMVStore.java
+++ b/h2/src/test/org/h2/test/store/TestMVStore.java
@@ -823,7 +823,7 @@ public class TestMVStore extends TestBase {
                                 " cache hit ratio: " + s.getFileStore().getCacheHitRatio() +
                                 " cache ToC hit ratio: " + s.getFileStore().getTocCacheHitRatio() +
                                 "",
-                        Math.abs(100 - (100 * expected / readCount)) < 15);
+                        Math.abs(100 - (100 * expected / readCount)) < 20);
             }
         }
     }

